src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java

```diff
@@ -79,17 +124,147 @@ enum Signal {
    */
   final Map<RepoRecordedInput, String> recordedInputValues = new TreeMap<>();
 
+  RepoFetchingSkyKeyComputeState(String repositoryName,
+      Function<RepoFetchingSkyKeyComputeState, Runnable> threadFunc) {
+    this.repositoryName = repositoryName;
+    workerThread =
+        Thread.ofVirtual()
+            .name("starlark-repository-" + repositoryName)
+            .start(threadFunc.apply(this));
+  }
+
   SkyFunction.Environment signalForFreshEnv() throws InterruptedException {
-    signalQueue.put(Signal.RESTART);
-    return delegateEnvQueue.take();
+    Response restartResponse = sendRequest(new Request.NewDependencies());
+    Preconditions.checkState(restartResponse instanceof Response.Restarting);
+    Response.Environment envResponse = (Response.Environment) sendRequest(new Request.GetEnvironment());
+    return envResponse.environment;
+  }
+
+  private void logMaybe(String msg) {
+    // System.err.println(msg);
   }
 
+  Response sendRequest(Request request) throws InterruptedException {
+    int seq = nextRequestSequenceNo++;
+    Packet<Request> packet = new Packet(seq, request);
+    logMaybe(String.format("LOG %s/worker: sending request %s", repositoryName, packet));
+    requestQueue.put(packet);
+    // This loop is necessary because the worker and host threads can get out of sync:
```

Comment:
As discussed offline, I think if we can start a new worker thread in case of `RestartDecision(true)`, so that the worker thread can terminate after 3), this loop can be removed. Anyway, I think this is also good.